### PR TITLE
Add custom transport tutorial

### PR DESCRIPTION
**Pull request by pablogs9**

### CODE REVIEW

File: _docs/tutorials/advanced/create_custom_transports/index.md [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+### Open function
+
+```c    
+bool my_custom_transport_open(uxrCustomTransport* transport)
+{
+    ...
+}
+```
+This function should open and init the custom transport. It returns a boolean indicating if the opening was successful. 
+`transport->args` have the arguments passed through `uxr_init_custom_transport`.
```
Comment (jamoralp) [inline _docs/tutorials/advanced/create_custom_transports/index.md:59]:
```suggestion
`transport->args` holds the arguments passed through `uxr_init_custom_transport`.
```

File: _docs/tutorials/advanced/create_custom_transports/index.md [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+### Close function
+```c    
+bool my_custom_transport_close(uxrCustomTransport* transport)
+{
+    ...
+}
+``` 
+This function should close the custom transport. It returns a boolean indicating if closing was successful. 
+`transport->args` have the arguments passed through `uxr_init_custom_transport`.
```
Comment (jamoralp) [inline _docs/tutorials/advanced/create_custom_transports/index.md:69]:
```suggestion
`transport->args` holds the arguments passed through `uxr_init_custom_transport`.
```

File: _docs/tutorials/advanced/create_custom_transports/index.md [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    const eprosima::uxr::CustomEndPoint* destination_endpoint,
+    uint8_t* buffer,
+    size_t length,
+    eprosima::uxr::TransportRc& transport_rc) -> ssize_t
+{
+    ...
+}
+```
+This function should write data to the custom transport. It must use
+the `destination_endpoint` members to set the data destination. It returns the number of Bytes written.
```
Comment (FranFin) [inline _docs/tutorials/advanced/create_custom_transports/index.md:180]:
```suggestion
the `destination_endpoint` members to set the data destination. It returns the number of bytes written.
```

File: _docs/tutorials/advanced/create_custom_transports/index.md [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        uxrCustomTransport* transport,
+        const uint8_t* buffer,
+        size_t length,
+        uint8_t* errcode)
+{
+    ...
+}
+```
+This function should write data to the custom transport. It returns the number of Bytes written.
+`transport->args` have the arguments passed through `uxr_init_custom_transport`.
```
Comment (jamoralp) [inline _docs/tutorials/advanced/create_custom_transports/index.md:83]:
```suggestion
`transport->args` holds the arguments passed through `uxr_init_custom_transport`.
```

File: _docs/tutorials/advanced/create_custom_transports/index.md [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+    ...
+}
+```
+This function should write data to the custom transport. It must use
+the `destination_endpoint` members to set the data destination. It returns the number of Bytes written.
+It should set `transport_rc` indicating the result of the operation.
+
+* **Stream-oriented mode:** The function can send up to `length` Bytes from `buffer`.
+
+* **Packet-oriented mode:** The function should send `length` Bytes from `buffer`. If less than `length` Bytes are written, `transport_rc` can be set.
```
Comment (FranFin) [inline _docs/tutorials/advanced/create_custom_transports/index.md:185]:
```suggestion
* **Packet-oriented mode:** The function should send `length` Bytes from `buffer`. If less than `length` bytes are written, `transport_rc` can be set.
```

File: _docs/tutorials/advanced/create_custom_transports/index.md [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+        uint8_t* buffer,
+        size_t length,
+        int timeout,
+        eprosima::uxr::TransportRc& transport_rc) -> ssize_t
+{
+    ...
+}
+```
+This function should read data to the custom transport. It must fill `source_endpoint` members with data source.
+It returns the number of Bytes read.
```
Comment (FranFin) [inline _docs/tutorials/advanced/create_custom_transports/index.md:200]:
```suggestion
It returns the number of bytes read.
```

File: _docs/tutorials/advanced/create_custom_transports/index.md [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+It should set `transport_rc` indicating the result of the operation.
```
Comment (FranFin) [inline _docs/tutorials/advanced/create_custom_transports/index.md:201]: Not totally clear what this means.

File: _docs/tutorials/advanced/create_custom_transports/index.md [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+
+* **Stream-oriented mode:** The function should retrieve up to `length` Bytes from transport
```
Comment (FranFin) [inline _docs/tutorials/advanced/create_custom_transports/index.md:203]:
```suggestion
* **Stream-oriented mode:** The function should retrieve up to `length` bytes from the transport
```

File: _docs/tutorials/advanced/create_custom_transports/index.md [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+    and write them into `buffer` in `timeout` milliseconds.
+
+* **Packet-oriented mode:** The function should retrieve `length` Bytes from transport
```
Comment (FranFin) [inline _docs/tutorials/advanced/create_custom_transports/index.md:206]:
```suggestion
* **Packet-oriented mode:** The function should retrieve `length` bytes from the transport
```

File: _docs/tutorials/advanced/create_custom_transports/index.md [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+    and write them into `buffer` in `timeout` milliseconds. If less than `length` Bytes are read transport_rc can be set.
```
Comment (FranFin) [inline _docs/tutorials/advanced/create_custom_transports/index.md:207]:
```suggestion
    and write them into `buffer` in `timeout` milliseconds. If less than `length` bytes are read, `transport_rc` can be set.
```

**Review by FranFin — CHANGES_REQUESTED** [review]
maybe clarify better usage of `errcode` in the Client and of `transport_rc` in the Agent.

File: _docs/tutorials/advanced/create_custom_transports/index.md [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+    my_custom_transport_open,
+    my_custom_transport_close,
+    my_custom_transport_write
+    my_custom_transport_read);
+
+custom_agent.start();
+```
+As in the *Client* API, four functions should be implemented. The behavior of these functions is sightly different
+depending on the selected mode.
+### CustomEndPoint
```
Comment (FranFin) [inline _docs/tutorials/advanced/create_custom_transports/index.md:141]:
```suggestion

### CustomEndPoint
```

File: _docs/tutorials/advanced/create_custom_transports/index.md [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+
+The `custom_endpoint` is an object of type `eprosima::uxr::CustomEndPoint` and it us in charge of handling the endpoint parameters. The *Agent*, unlike the *Client*, can receive
```
Comment (FranFin) [inline _docs/tutorials/advanced/create_custom_transports/index.md:143]:
```suggestion
The `custom_endpoint` is an object of type `eprosima::uxr::CustomEndPoint` and it is in charge of handling the endpoint parameters. The *Agent*, unlike the *Client*, can receive
```

File: _docs/tutorials/advanced/create_custom_transports/index.md [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+messages from multiple *Clients* so it must be able to differentiate between different *Clients*.
```
Comment (FranFin) [inline _docs/tutorials/advanced/create_custom_transports/index.md:144]:
```suggestion
messages from multiple *Clients* so it must be able to differentiate between them.
```

File: _docs/tutorials/advanced/create_custom_transports/index.md [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+Therefore, the `eprosima::uxr::CustomEndPoint` should be provided with information about the origin of the message
+in the read callback, and with information about the destination of the message in the write callback.
+
+In general, members of a `eprosima::uxr::CustomEndPoint` object can be unsigned integers and strings.
```
Comment (FranFin) [inline _docs/tutorials/advanced/create_custom_transports/index.md:148]:
```suggestion
In general, the members of a `eprosima::uxr::CustomEndPoint` object can be unsigned integers and strings.
```

File: _docs/tutorials/advanced/create_custom_transports/index.md [hunk 15]
```diff
@@ -0,0 +1,240 @@
...
+
+`CustomEndPoint` defines three methods:
+
+Add member
+```cpp
+bool eprosima::uxr::CustomEndPoint::add_member<*KIND*>(const std::string& member_name);
+```
+Allows to dynamically add a new member to the endpoint definition.
```
Comment (FranFin) [inline _docs/tutorials/advanced/create_custom_transports/index.md:156]:
```suggestion
This function allows to dynamically add a new member to the endpoint definition.
```

File: _docs/tutorials/advanced/create_custom_transports/index.md [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+
+Returns `true` if member was correctly added, or `false` if something went wrong (for example, the member already existed).
```
Comment (FranFin) [inline _docs/tutorials/advanced/create_custom_transports/index.md:158]:
```suggestion
Ir returns ``true`` if the member was correctly added, ``false`` if something went wrong (for example, if the member already exists).
```

File: _docs/tutorials/advanced/create_custom_transports/index.md [hunk 17]
```diff
@@ -0,0 +1,240 @@
...
+
+- **KIND**: To be chosen from: `uint8_t`, `uint16_t`, `uint32_t`, `uint64_t`, `uint128_t` or `std::string`.
+- **member_name**: The tag used to identify the endpoint member.
+
+Set member value
+```cpp
+void eprosima::uxr::CustomEndPoint::set_member_value(const std::string& member_name, const *KIND* & value);
+```
+
+Sets the specific value (numeric or string) for a certain member, which must previously exist in the CustomEndPoint.
```
Comment (FranFin) [inline _docs/tutorials/advanced/create_custom_transports/index.md:168]:
```suggestion
This function sets the specific value (numeric or string) for a certain member, which must previously exist in the `CustomEndPoint`.
```

File: _docs/tutorials/advanced/create_custom_transports/index.md [hunk 18]
```diff
@@ -0,0 +1,240 @@
...
+
+- **member_name**: The member whose value is going to be modified.
+- **value**: The value to be set, of `KIND`: `uint8_t`, `uint16_t`, `uint32_t`, `uint64_t`, `uint128_t` or `std::string`.
+
+Get member
+```cpp
+const *KIND* & eprosima::uxr::CustomEndPoint::get_member(const std::string& member_name);
+```
+
+Gets the current value of the member registered with the given parameter.
```
Comment (FranFin) [inline _docs/tutorials/advanced/create_custom_transports/index.md:178]:
```suggestion
This function gets the current value of the member registered with the given parameter.
```

**Review by FranFin — CHANGES_REQUESTED** [review]

**Review by FranFin — APPROVED** [review]

**Review by jamoralp — APPROVED** [review]

**Pull request merged**